src/twisted/web/test/test_web.py

```diff
@@ -751,6 +751,41 @@ def test_retrieveNonExistentSession(self):
         self.assertNotIn(b"does-not-exist", request.cookies[0])
 
 
+    def test_getSessionExpired(self):
+        """
+        L{Request.getSession} generates a new session when the previous
+        session has expired.
+        """
+        clock = Clock()
+        site = server.Site(resource.Resource())
+        d = DummyChannel()
+        request = server.Request(d, 1)
+        request.site = site
+        request.sitepath = []
+
+        def SessionFactoryWithClock(site, uid):
+            """
+            Forward to normal session factory, but inject the clock.
+            """
+            session = OriginalSessionFactory(site, uid)
+            session._reactor = clock
+            return session
+
+        # The site is patch to allow injecting a clock to the session.
+        OriginalSessionFactory = site.sessionFactory
+        site.sessionFactory = SessionFactoryWithClock
+
+        initialSession = request.getSession()
+
+        # When the session is requested after the session timeout,
+        # no error is raised and a new session is returned.
+        clock.advance(OriginalSessionFactory.sessionTimeout )
+        newSession = request.getSession()
+
+        self.assertIsNot(initialSession, newSession)
```

Comment:
This is the most important question for the whole PR ... what should be true about the result of `getSession` after a prior session has expired?

The possibilities:

1. You get back the same Session object as you had before, but with its lifetime extended as though you've just `touch()`'d it.
2. You get back a new Session object but with the same uid as the one you had before with a life time of `sessionTimeout` starting at the point of the `getSession` call.
3. You get back a new Session object with a different uid - basically a whole new session, as if the previous one never existed.

Option 3 makes the most sense to me.  Once the old session expires it seems like all remaining behavior should be as though it never existed in the first place.  But ...

If you are handling a request using two libraries, call them A and B, and near the beginning of handling the request library A asks for the session (creating it, perhaps) and near the end of handling the request library B asks for the session, then what is the desired behavior if enough time separates those two events sufficiently for the session to expire?  Do you _want_ to be able to get two different sessions while handling a single request?

It seems like it might be even more reasonable, on a particular request, the session could never exist and then not exist.  Either the request is associated with a session or it is not.  It shouldn't be associated with two sessions, should it?

So perhaps I will propose that instead this assertion should be exactly flipped?  `assertIs(initialSession, newSession)`?

What do you think?  Does this example scenario sound compelling?  Does it differ from any real-world cases you've encountered?